### ISSUE TITLE
fit_generator Multiprocessing Overhead - Performance Optimization Ideas?

### ISSUE THREAD

**Issue by mharradon** [keras-team/keras#11578]
I'm training a model using fit_generator with multiprocessing. I've been working on optimizing training performance, and I've found via some profiling that roughly half of the time training is spent in various multiprocessing _recv methods (the other half waiting on tensorflow for GPU callbacks). It appears that significant time is spent transferring bytes and pickling/unpickling (I just benchmarked, and the pickling/unpickling is sufficient to explain the time).

My workload is moderately extreme in terms of batch size and number of GPUs, but I would imagine many people are unknowingly getting limited by this. I was curious if anyone had suggestions or if there was interest in contributions here - I'm considering implementing something to improve this situation, though I'm not sure what yet.

**Comment by gabrieldemarmiesse** [comment]
I suppose that would be something to discuss with @Dref360 . It's not really my area of expertise.

**Comment by Dref360** [comment]
Not sure how to speed this up. If you could help that would be amazing :D 

We send the Sequence object once per epoch at each worker. (We need to modify the Sequence according to `on_epoch_end` and then recreate the pool with the updated version.
 
Then each worker computes __getitem__ and returns the batch. (This is where the IO bottleneck happens)

Maybe they could store the batch in a shared memory space, but we would need to know the exact size of a batch which is troublesome. 

Have you tried with "large" batches? Some that take 1-2 seconds to compute to see if you get similar behavior.

**Comment by mharradon** [comment]
Currently I have 2 ideas:

1. Some sort of shared memory approach like you suggested. I'm thinking an additional process with 2 shared memory buffers that flip (like double buffering in graphics). One process deserializes from the Queue to some shared memory space while the main process runs on the GPU, then the shared memory buffers flip and it can read the previously deserialized data. The fixed size may be a limitation though.
2. Improved serialization/deserialization. Using numpy `frombuffer` and `tobytes` is roughly 3-8x faster than pickling in my tests. I was hoping that I could wrap `np.ndarray` objects and provide simpler `__getstate__` and `__setstate__` methods, though so far that doesn't seem to help much (x2 at best, more like x1.5). It may require a modified Queue/multiprocessing implementation that doesn't use cPickle (specifically pickle.loads appears to be the slow spot). That would probably be a pain from a maintenance and distribution perspective...

>Have you tried with "large" batches?

My batches are ~500MB. They take a long time to compute, so I have a bunch of workers and a large buffer. In profiling (via [py-spy](https://github.com/benfred/py-spy), amazing tool) I see that the tensorflow session calls are ~half the code run time. Training in total is running at around 3-4s/batch. `pickle.loads` takes around 2s on arrays of that size, so that currently appears to be the pain point for me. It's certainly possible I'm pushing up against memory copying speed limitations on my hardware as well.

**Comment by mharradon** [comment]
I've got a preliminary double-buffering shared memory implementation up and in some preliminary measurements on my workload I'm seeing ~2x speedup in total training (!). I'm monitoring that the queue is full, so I don't think that was an issue.

One thing I've already noticed is that (for my workload) double-buffering isn't quite enough - I need a few buffers of shared memory to keep up with the multiple GPUs I'm using, so it's basically a second queue. This might indicate that a more long-term approach would be to migrate the core queue system to shared memory when possible. It could always fall back to messaging if needed, and it would even be feasible to create expanding buffers to accommodate variably sized batches.

My expectation is that many multigpu systems (or systems using the Titan Vs or 2080Tis) could see a performance hit here, so the complexity in implementations might be worth a hopefully wide-ranging performance improvement.

**Comment by mharradon** [comment]
@Dref360, would an implementation along these lines be of interest? I've got most of a solution that reallocates shared memory buffers to match the largest observed batch sizes. The batches are packed into raw bytes and written to the SHM buffer, then the training process slices them apart into lists of arrays for each x and y (w/ correct dtype as well) in constant time by pointing to the start in the SHM buffer. If batch size is too large for the existing SHM buffer I currently fall back to passing through a pipe, so asymptotically you should get the full speedup even for variably sized batches and worst case is equivalent to the current performance. Reallocation of SHMs to replace undersized SHMs has to happen on the host process, but that appears fast (and should eventually not happen as training proceeds).

It's a moderately complex arrangement to handle the dynamic sizes - of course fixed size is comparatively straight-forward. The potentially significant performance gains seem worth it, and I'd expect the advantage to increase over time as typical GPU FLOPS /(memory bandwidth) increases.

**Comment by Dref360** [comment]
I think this would be a valuable addition as long as the users see little/no impact. 
If the code is quite large, maybe it's time to have a keras.utils.multiprocessing package.

Some common *gotcha* we have to handle:
* tests are sometime disabled for multiprocessing, please be sure that they are runned.
* Should work on Windows and UNIX (fork vs spawn)
* Should work on Python 2 and 3
* Multiple Enqueuers should be able to work at the same time (Training/validation)
* We should avoid adding any new dependency
* Clean-up on `close`

**Comment by jvishnuvardhan** [comment]
@ is there any issue or you are satisfied with the performance? Please close if the issue was resolved by updating keras. thanks!
